vllm/model_executor/layers/quantized_linear/gptq.py

```diff
@@ -0,0 +1,185 @@
+from torch.nn.parameter import Parameter
+
+from vllm import quantization_ops
+from vllm.model_executor.parallel_utils.layers import (ColumnParallelLinear,
+                                                       RowParallelLinear)
+
+
+# TODO(woosuk): Optimize.
+def _gptq_matmul(
+    x: torch.Tensor,
+    qweight: torch.Tensor,
+    qzeros: torch.Tensor,
+    scales: torch.Tensor,
+    g_idx: torch.Tensor,
+    shifter: torch.Tensor,
+) -> torch.Tensor:
+    """Matrix multiplication with GPTQ weights."""
+    # qw: [input_size, output_size]
+    qw = (qweight.unsqueeze(1) >> shifter.view(1, -1, 1)) & 0xf
+    qw = qw.flatten(start_dim=0, end_dim=1)
+
+    # qz: [input_size, output_size]
+    qz = (qzeros[g_idx].unsqueeze(2) >> shifter.view(1, 1, -1)) & 0xf
+    qz = qz + 1
+    qz = qz.flatten(start_dim=1, end_dim=2)
+
+    # qs: [input_size, output_size]
+    qs = scales[g_idx]
+    # w: [input_size, output_size]
+    w = qs * (qw - qz).to(qs.dtype)
+
+    # out: [batch_size, output_size]
+    out = torch.matmul(x, w)
+    return out
+
+
+class GPTQColumnParallelLinear(ColumnParallelLinear):
+
+    def create_weights(self, dtype: torch.dtype) -> None:
+        assert self.input_size % self.quant_config.pack_factor == 0
+        assert self.input_size % self.quant_config.group_size == 0
+        assert (self.output_size_per_partition %
+                self.quant_config.pack_factor == 0)
+
+        group_size = self.quant_config.group_size
+        if group_size == -1:
+            group_size = self.input_size
+
+        self.qweight = Parameter(
+            torch.empty(
+                self.input_size // self.quant_config.pack_factor,
+                self.output_size_per_partition,
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        self.qzeros = Parameter(
+            torch.empty(
+                self.input_size // group_size,
+                self.output_size_per_partition //
+                self.quant_config.pack_factor,
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        self.scales = Parameter(
+            torch.empty(
+                self.input_size // group_size,
+                self.output_size_per_partition,
+                device="cuda",
+                dtype=dtype,
+            ),
+            requires_grad=False,
+        )
+        # Initialize g_idx to be sequential.
+        # This is required because old GPTQ models may not have g_idx.
+        self.g_idx = Parameter(
+            torch.tensor(
+                [i // group_size for i in range(self.input_size)],
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        self.shifter = torch.tensor(
+            [0, 4, 8, 12, 16, 20, 24, 28],
+            device="cuda",
+            dtype=torch.int32,
+        )
```

Comment:
Is this `shifter` a must-have? The problem here is that we created a "non-parameter" tensor. We will need to modify the weight creation code in #1622 to make this creation work.

Reply:
How difficult will it be to add support for such tensors? While this tensor is only used for the pytorch-based GPTQ matmul implementation and thus will be eventually unused once we develop a more optimized kernel, such non-parameter buffers can be used for other quantization methods. I believe we should take this into account in the new design.